Port typst info to tytanic util about

T1mVo opened the pull request:
**Typst:**

<img width="616" height="614" alt="image" src="https://github.com/user-attachments/assets/f958e11f-2602-4c06-9d6e-195bda495c79" />

**Tytanic:**

<img width="616" height="443" alt="image" src="https://github.com/user-attachments/assets/2396e904-2914-4826-b033-e7820987dde3" />

Resolves #228
Resolves #262

T1mVo committed:
Add compile time retrieval of build information

Retrieves the following build information at compile time:
- The version of tytanic
- The version of typst
- The current commit hash

Can be manually overwritten by using the following envs:
- `TYTANIC_VERSION`
- `TYTANIC_TYPST_VERSION`
- `TYTANIC_COMMIT_SHA`

T1mVo committed:
Add compile time retrieval of build information

Retrieves the following build information at compile time:
- The version of tytanic
- The version of typst
- The current commit hash

Can be manually overwritten by using the following envs:
- `TYTANIC_VERSION`
- `TYTANIC_TYPST_VERSION`
- `TYTANIC_COMMIT_SHA`

T1mVo committed:
Edit `tt util about` command

This is a vague port of the `typst info` command.
It shows build informations, font informations, package informations
and environment informations.

T1mVo committed:
Edit `tt util about` command

This is a vague port of the `typst info` command.
It shows build informations, font informations, package informations
and environment informations.

tingerrr commented on crates/tytanic/src/cli/commands/util/about.rs:
```diff
@@ -1,13 +1,229 @@
-use std::io::Write;
+use std::{env, io::Write, path::PathBuf};
```
Could you split these into individual lines? Unfortunately, this is still a nightly rustfmt feature, so your rust-analyzer likely rejected the config in the repo.

tingerrr commented on crates/tytanic/src/cli/commands/util/about.rs:
```diff
@@ -1,13 +1,229 @@
 use color_eyre::eyre;
+use termcolor::Color;
+
+use crate::{cli::commands::Switch, cwrite, cwriteln};
```
Same as above.

tingerrr commented on crates/tytanic/src/cli/commands/util/about.rs:
```diff
@@ -1,13 +1,229 @@
+        let package_cache_path = match &ctx.args.package.package_cache_path {
+            Some(package_cache_path) => Some(package_cache_path.clone()),
+            None => typst_kit::package::default_package_cache_path(),
+        };
```
Yes, we use the same paths here.

tingerrr commented on crates/tytanic/src/cli/commands/util/about.rs:
```diff
@@ -1,13 +1,229 @@
+    }
+
+    cwrite!(colored(w, Color::Cyan), "  System fonts   ")?;
+    cwriteln!(
```
It would be great if the `on`/`off` states are aligned with the system fonts above.

In the original Typst implementation I wrote a small key-value formatting abstraction to take care of this, I think you can copy that almost verbatim.

T1mVo replied:
I just aligned it manually for now as I was having trouble with the `cwrite` macro and passing the writer into functions.

tingerrr commented on crates/tytanic/src/cli/commands/util/about.rs:
```diff
@@ -1,13 +1,229 @@
+    /// The version of tytanic.
+    version: &'static str,
+    /// The typst version used in this instance of tytanic.
+    typst_version: &'static str,
```
Could you move both of the versions to `About`?

tingerrr reviewed (changes requested):
Looks great! I've left a few comments, mostly in relation to consistency with Typst.

The structs should be made serializable and a `--json` option should be added such that this can be used in a script.

Since this basically re-implements the logic I wrote for Typst, it may be appropriate to extract parts of this into `typst-kit` at some point before the next Typst release so we avoid having stale information here. But for now I'd go ahead and merge it after you addressed the other comments.

Thanks!

T1mVo committed:
Format with nightly toolchain

T1mVo committed:
Format with nightly toolchain

tingerrr commented on crates/tytanic/build.rs:
```diff
@@ -0,0 +1,75 @@
+        return Cow::Borrowed(version);
+    }
+
+    let cargo_version = Command::new("cargo")
```
No, I specifically meant `cargo release --dry-run` because this will run on a new directory, and I'm unsure if all the necessary information is available. I can try it later today.

T1mVo replied:
Please test this. I dont have cargo release installed.

tingerrr replied:
Seems to work. :+1:

T1mVo committed:
Rename commit_sha to commit

Co-authored-by: Erik <tinger@tinger.dev>

T1mVo committed:
Rename commit_sha to commit

Co-authored-by: Erik <tinger@tinger.dev>

T1mVo committed:
Move versions to About

T1mVo committed:
Move versions to About

T1mVo committed:
Add platform struct

T1mVo committed:
Add platform struct

T1mVo committed:
Add json serialization

T1mVo committed:
Add json serialization

T1mVo committed:
Add further env vars

T1mVo committed:
Add further env vars

T1mVo committed:
Add json and yaml output

T1mVo committed:
Add json and yaml output

T1mVo committed:
Change coloring in output

T1mVo committed:
Change coloring in output

T1mVo committed:
cli: Change parsing of tytanic commit sha

If the commit sha cannot be determined, no environment variable will be set.
The "unknown commit" message is handled in `about.rs`.

T1mVo committed:
cli: Reorder env vars in util about

The order matches with Typst as of version 0.15.

T1mVo committed:
cli: Change alignment of font settings in util about

Matches the alignment of Typst as of 0.15.

T1mVo committed:
cli: Fix typos and inconsistencies
